pkg/uptimerobot.go

```diff
@@ -114,15 +114,15 @@ func (a AlertContact) String() string {
 
 // Monitor represents an UptimeRobot monitor.
 type Monitor struct {
-	ID           int64  `json:"id"`
+	ID           int64  `json:"id,omitempty"`
 	FriendlyName string `json:"friendly_name"`
 	URL          string `json:"url"`
 	Type         int    `json:"type"`
 	// keyword_type, sub_type, and port are returned as either an integer
 	// (if set) or an empty string (if unset), which Go's JSON library won't
 	// parse for integer fields: https://github.com/golang/go/issues/22182
-	SubType       interface{} `json:"sub_type"`
-	KeywordType   interface{} `json:"keyword_type"`
+	SubType       interface{} `json:"sub_type,omitempty"`
+	KeywordType   interface{} `json:"keyword_type,omitempty"`
```

Comment:
Why is this an interface type?

vs., using an enum like:

https://github.com/bitfield/uptimerobot/pull/13/files#diff-5f6ef36e76a0ffc232cb50f0bb90e2e6R115

Then https://github.com/bitfield/uptimerobot/pull/12/files#diff-5f6ef36e76a0ffc232cb50f0bb90e2e6R403 wouldn't be necessary right?

Reply:
Ah, I see your comment, https://github.com/bitfield/uptimerobot/blob/f0c2884051c7f9b52eba410854a5c0cc9d7cea81/pkg/uptimerobot.go#L121

Reply:
I'm actually confused by the comment. The bug that it links to only occurs if the JSON has an integer stored inside a string like this:

```json
{"sub_type": "1"}
```

I don't believe the uptimerobot API does that, so

```
SubType       int `json:"sub_type"`
```

should be fine I think..

Reply:
It's getting late. I just realized that it still fails. For what it's worth, I had to write some code sometime where I wanted to deserialize some YAML into a custom type (time.Duration) from a string "30m".

Something similar can be done here, https://play.golang.org/p/mmjr5hktU-h

Reply:
You put your finger accurately on my biggest current bugbear with the library. The `interface{}` is basically a nasty hack to work around a bug in the Uptime Robot API (which they have been informed about, but I don't think it's yet been fixed). If `sub_type` is null, the API returns an empty string rather than, for example, 0. This causes the Go JSON parser to choke if the field is of type `int`.

I've learned quite a bit about writing API clients since doing this one (principally, by doing another one), and I'm now ready to apply some of those lessons back to this one. As part of that I want to iron out this issue, and do several other refactorings. The first step before that, though, is to get some decent tests in.

If you have a look at [checkly](github.com/bitfield/checkly), this is a whole lot nicer in many ways, and this is how I'd like the Uptime Robot client to end up looking.

Reply:
Some big improvements to the tests (which could still do with actually testing much more than they do, but that'll come). I'll work on fixing the JSON issues next.